UCLALibrary/LBS
# SYS-1415: Import Excel data

Implements [SYS-1415: Import Excel data](https://uclalibrary.atlassian.net/browse/SYS-1415) and [SYS-1412: Create upload form/view](https://uclalibrary.atlassian.net/browse/SYS-1412), as these are closely related.  
Internal only, not tagged for deployment.

This PR adds the ability to import data from the 3 Excel files LBS provides.  Each file is read, data is mapped to the appropriate model, and each row of data is saved as an instance of that model.  All existing data for these models is deleted before new data is saved, so the contents of each Excel file replace previous data.

Other changes / notes:
* `pandas` package is added to support loading Excel data, along with the `xlrd` package `pandas` uses internally (but does not explicitly require for installation) to read older `.xls` Excel file.
* Models added in #74 are all updated to allow all columns to be `Null`.  There's no data validation of this imported data.
* `ge.views_utils.get_mapping()` defines the column mappings between the Excel files and the models.
* There's a very basic template to display the upload form; styling will come later in the project.
* There are 3 sample files, each with a few rows of data from the real Excel files LBS provided.  These are used by the automated tests, and can be used for manual testing:
```
sample_files/
├── sample_consolidated.xls
├── sample_ledger.xls
└── sample_mtf.xls
```

### Testing
Restart the application to install new requirements and to apply migrations.
Run tests.  There are 6 for the `ge` app, and 35 (total) for the whole project; all should pass:
`docker-compose exec django python manage.py test ge`
`docker-compose exec django python manage.py test`

* Log into http://127.0.0.1:8000 as admin/admin
* Then go manually to http://127.0.0.1:8000/ge/report/ since I haven't rewired authentication URLs yet...
* Upload files, either samples or full ones from LBS
  * BFS -> consolidated
  * CDW -> ledger
  * MTF -> mtf
* If files were correctly chosen, a generic success message should appear.
* If any files are mismatched (e.g., BFS got the ledger file), an error message should appear, like `"Unexpected column 'Source' in sample_ledger.xls for BFSImport"`
* Django should require a file be selected for all 3 files.
